httpmw/logging/logging.go

```diff
@@ -52,22 +54,61 @@ func (l *loggingMiddleware) ServeHTTP(w http.ResponseWriter, r *http.Request) {
 	customWriter := &responseWriter{ResponseWriter: w}
 
 	startTime := l.now()
-	l.LogRequestStart(r)
-	l.next.ServeHTTP(customWriter, r)
+	ctx := r.Context()
+
+	l.LogRequestStart(ctx, r)
+	l.next.ServeHTTP(customWriter, r.WithContext(ctx))
 	endTime := l.now()
 
 	latency := endTime.Sub(startTime)
-	l.LogRequestEnd(r, "RequestEnd", customWriter.statusCode, latency)
-	l.LogRequestEnd(r, "finished call", customWriter.statusCode, latency)
+	l.LogRequestEnd(ctx, r, "RequestEnd", customWriter.statusCode, latency)
+	l.LogRequestEnd(ctx, r, "finished call", customWriter.statusCode, latency)
 }
 
-func (l *loggingMiddleware) LogRequestStart(r *http.Request) {
-	l.logger.Info("RequestStart", "source", "ApiRequestLog", "protocol", "HTTP", "method_type", "unary",
-		"component", "client", "method", r.Method, "service", r.Host, "url", r.URL.String())
+func (l *loggingMiddleware) LogRequestStart(ctx context.Context, r *http.Request) {
+	md, ok := metadata.FromIncomingContext(ctx)
+	attributes := []interface{}{
+		"source", "ApiRequestLog",
+		"protocol", "HTTP",
+		"method_type", "unary",
+		"component", "client",
+		"method", r.Method,
+		"service", r.Host,
+		"url", r.URL.String(),
+	}
+
+	if ok {
+		for key, values := range md {
+			for _, value := range values {
+				attributes = append(attributes, key, value)
+			}
+		}
+	}
+
+	l.logger.InfoContext(ctx, "RequestStart", attributes...)
 }
 
-func (l *loggingMiddleware) LogRequestEnd(r *http.Request, msg string, statusCode int, duration time.Duration) {
-	l.logger.Info(msg, "source", "ApiRequestLog", "protocol", "HTTP", "method_type",
-		"unary", "component", "client", "method", r.Method, "service", r.Host,
-		"url", r.URL.String(), "code", statusCode, "time_ms", duration.Milliseconds())
+func (l *loggingMiddleware) LogRequestEnd(ctx context.Context, r *http.Request, msg string, statusCode int, duration time.Duration) {
```

Comment:
Merge to be one function. It is hard to maintain similar/duplicate code.

Reply:
The reason we currently have two functions is to remain consistent with aks-rp. Nick mentioned users need to be able to call a RequestStart and RequestEnd function for logging/alert purposes

Reply:
What I mean is these two functions have many duplicate parts. You can either make it to be one function (adding a parameter to tell if it is for start or for end) or extract the duplicate part as a helper function.

Avoid duplicate code!

Reply:
using buildAttributes() helper function to extract out common logic